mixing/mixpool/mixpool_test.go

```diff
@@ -0,0 +1,399 @@
+func (m *mockUTXO) PkScript() []byte      { return m.output.PkScript }
+func (m *mockUTXO) ScriptVersion() uint16 { return 0 }
+func (m *mockUTXO) BlockHeight() int64    { return m.blockHeight }
+func (m *mockUTXO) Amount() int64         { return m.output.Value }
+
+func makeMockUTXO(rand io.Reader, value int64) *mockUTXO {
+	tx := wire.NewMsgTx()
+
+	pub, priv, err := generateSecp256k1(rand)
+	if err != nil {
+		panic(err)
+	}
+	pubSerialized := pub.SerializeCompressed()
+
+	p2pkhScript := []byte{
+		0:  txscript.OP_DUP,
+		1:  txscript.OP_HASH160,
+		2:  txscript.OP_DATA_20,
+		23: txscript.OP_EQUALVERIFY,
+		24: txscript.OP_CHECKSIG,
+	}
+	hash160 := dcrutil.Hash160(pubSerialized)
+	copy(p2pkhScript[3:23], hash160)
+
+	output := wire.NewTxOut(value, p2pkhScript)
+	tx.AddTxOut(output)
+
+	return &mockUTXO{
+		outpoint: wire.OutPoint{
+			Hash:  tx.TxHash(),
+			Index: 0,
+			Tree:  0,
+		},
+		tx:      tx,
+		output:  output,
+		pubkey:  pubSerialized,
+		privkey: priv,
+	}
+}
+
+func (m *mockUTXO) mixprutxo(expires uint32) wire.MixPairReqUTXO {
+	k := utxoproof.Secp256k1KeyPair{
+		Pub:  m.pubkey,
+		Priv: m.privkey,
+	}
+	sig, err := k.SignUtxoProof(expires)
+	if err != nil {
+		panic(err)
+	}
+
+	return wire.MixPairReqUTXO{
+		OutPoint:  m.outpoint,
+		PubKey:    m.pubkey,
+		Signature: sig,
+	}
+}
+
+func makeMockUTXOs(rand io.Reader) map[string]*mockUTXO {
+	return map[string]*mockUTXO{
+		"A": makeMockUTXO(rand, 20e8),
+	}
+}
+
+type fakechain struct {
+	hash   chainhash.Hash
+	height int64
+}
+
+func (c *fakechain) ChainParams() *chaincfg.Params {
+	return params
+}
+
+func (c *fakechain) CurrentTip() (chainhash.Hash, int64) {
+	return c.hash, c.height
+}
+
+func (c *fakechain) FetchUtxoEntry(op wire.OutPoint) (UtxoEntry, error) {
+	entry := utxoStore.byOutpoint[op]
+	if entry == nil {
+		return nil, errors.New("no utxo entry")
+	}
+	return entry, nil
+}
+
+func TestAccept(t *testing.T) {
+	t.Parallel()
+	testRand := testPRNG(t)
+
+	c := new(fakechain)
+	c.height = 1000
+	p := NewPool(c)
+
+	identityPub, identityPriv, err := generateSecp256k1(testRand)
+	if err != nil {
+		t.Fatal(err)
+	}
+	identity := *(*[33]byte)(identityPub.SerializeCompressed())
+
+	var (
+		expires      uint32 = 1010
+		mixAmount    int64  = 10e8
+		scriptClass         = mixing.ScriptClassP2PKHv0
+		txVersion    uint16 = wire.TxVersion
+		lockTime     uint32 = 0
+		messageCount uint32 = 1
+		inputValue   int64  = 20e8
+		utxos        []wire.MixPairReqUTXO
+		change       *wire.TxOut
+		flags        byte = 0
+		pairingFlags byte = 0
+	)
+	utxos = []wire.MixPairReqUTXO{
+		utxoStore.byName["A"].mixprutxo(expires),
+	}
+	pr, err := wire.NewMsgMixPairReq(identity, expires, mixAmount,
+		string(scriptClass), txVersion, lockTime, messageCount,
+		inputValue, utxos, change, flags, pairingFlags)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = mixing.SignMessage(pr, identityPriv)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = p.AcceptMessage(pr)
```

Comment:
This test code seems incorrect.  It creates a new message, but never calls `WriteHash` which means later when `AcceptMessage` calls `pr.Hash()` it's all zeros.  Then, later, the same thing applies to the remaining `KE`, `CT`, `SR`, `DC-net`, and `CM` messages which means the pool thinks it already has the messages due to them all having hashes of all zeros and never accepting them.

The `mempool` tests have assertions after every call to check the expected counts of items to help prevent this type of thing.

It seems like a pretty major design flow with the wire types tbh.  Callers having to preemptively call `WriteHash` on new messages is a major footgun as evidenced here.

Reply:
Your feedback is noted, but this test was written prior to that change in wire.  Earlier versions used to create the hash on the fly each call to Hash().  It's probably more worrying that everything keeps working despite all the zero hashes.  Perhaps a check in AcceptMessage for zero hashes would be ideal, to prevent this possible situation.